openfisca_france/model/prelevements_obligatoires/impot_revenu/charges_deductibles.py

```diff
@@ -681,10 +708,100 @@ def formula_2009(foyer_fiscal, period, parameters):
         '''
         Dépenses de grosses réparations des nus-propriétaires (case 6CB et 6HJ)
         '''
+
+        f6cb = foyer_fiscal('f6cb', period)
```

Comment:
Le code de cette formule me parait être beaucoup de copié-collé. Est-ce nécéssaire ?
Quel serait le risque de faire : 
```py
class grosses_reparations(Variable):
    value_type = float
    entity = FoyerFiscal
    label = u"Dépenses de grosses réparations des nus-propriétaires"
    definition_period = YEAR

    def formula_2009(foyer_fiscal, period, parameters):
        '''
        Dépenses de grosses réparations des nus-propriétaires
        '''
        somme_reparations = (
               foyer_fiscal('f6cb', period) +
               foyer_fiscal('f6cj', period) + # report année 2009
               foyer_fiscal('f6hk', period) +  # report année 2010
               foyer_fiscal('f6hl', period) +  # report année 2011
               foyer_fiscal('f6hm', period) +  # report année 2012
               foyer_fiscal('f6hn', period) +  # report année 2013
               foyer_fiscal('f6ho', period) +  # report année 2014
               foyer_fiscal('f6hp', period)   # report année 2015
        )
   
        plafond_grosses_reparations = parameters(period).impot_revenu.charges_deductibles.grosses_reparations.plafond

        return min_( somme_reparations, plafond_grosses_reparations)
```

De ce que j'ai compris, `f6hj` n'existe qu'à partir de 2010, `f6hk` qu'à partir de 2011 ...
Comme ce sont des inputs variables, elle n'ont pas de start date définie, ce qui voudrait dire qu'un utilisateur pourrait par erreur les renseigner sur une période antérieure à leur création, ce qui rendrait le calcul faux.

Or dans OF, c'est à l'utilisateur d'envoyer les bons inputs. Même lorsqu'une variable a une `end_date`, elle sera supplantée par un input mis par l'utilisateur.
ex : si je met une valeur en `2018-01` à la `paje_clca`(dont le `end_date` est `2017-04`), le calcul de la `paje` pour `2018-01` sera affecté.

Si vous estimez que le risque est trop important, il est possible d'écrire : 
```py
    def formula_2009(foyer_fiscal, period, parameters):
        '''
        Dépenses de grosses réparations des nus-propriétaires
        '''
            somme_reparations = (
                        foyer_fiscal('f6cb', period) +
                        np.where(period > '2009', foyer_fiscal('f6hj', period)) +  # report année 2009
                        np.where(period > '2010', foyer_fiscal('f6hk', period))  # report année 2010
                        np.where(period > '2011', foyer_fiscal('f6hl', period)) # report année 2011
                        np.where(period > '2012', foyer_fiscal('f6hm', period)) # report année 2012
                        np.where(period > '2013', foyer_fiscal('f6hn', period)) # report année 2013
                        np.where(period > '2014', foyer_fiscal('f6ho', period)) # report année 2014
                        np.where(period > '2015', foyer_fiscal('f6hp', period)) # report année 2015
        )

        plafond_grosses_reparations = parameters(period).impot_revenu.charges_deductibles.grosses_reparations.plafond

        return min_(somme_reparations, plafond_grosses_reparations)
```

Reply:
@Anna-Livia : Il y avait des if sur les period il y a un certain temps et il avait décidé si je me souviens bien de plutôt choisir des formules par périodes.
L'idée étant qu'il fallait simplifier le code pour l'utilisateur qui le lirait à une date donnée dans le legislation-explorer quitte à ne pas factoriser au maximum.

Reply:
@benjello C'est dans cet idée que je propose la première option, qui est à mon avis au plus proche de l'esprit d'OpenFisca. 
Les + : 
- Plus court donc plus lisible
- pas de copié/collé de code donc plus maintenable
- Respecte le fonctionnement global des inputs utilisateurs qui prennent le pas sur les formules.

Le moins : 
- risque de mauvais résultats si les periodes des inputs ne sont pas respecté par les utilisateurs. Ce qui est le risque de n'importe quelle variable (input ou formule) d'OpenFisca

Ma préconisation est donc :
```py
class grosses_reparations(Variable):
    value_type = float
    entity = FoyerFiscal
    label = u"Dépenses de grosses réparations des nus-propriétaires"
    definition_period = YEAR

    def formula_2009(foyer_fiscal, period, parameters):
        '''
        Dépenses de grosses réparations des nus-propriétaires
        '''
        somme_reparations = (
               foyer_fiscal('f6cb', period) +
               foyer_fiscal('f6cj', period) + # report année 2009
               foyer_fiscal('f6hk', period) +  # report année 2010
               foyer_fiscal('f6hl', period) +  # report année 2011
               foyer_fiscal('f6hm', period) +  # report année 2012
               foyer_fiscal('f6hn', period) +  # report année 2013
               foyer_fiscal('f6ho', period) +  # report année 2014
               foyer_fiscal('f6hp', period)   # report année 2015
        )
   
        plafond_grosses_reparations = parameters(period).impot_revenu.charges_deductibles.grosses_reparations.plafond

        return min_( somme_reparations, plafond_grosses_reparations)
```

Reply:
@Anna-Livia : je me permets de contester tes +
  - Plus cour donc plus lisible _is in the eye of the beholder_: si je veux regarder la formule pour une année il est plus lisible pour moi de l'avoir en entier séparément. Certes cela fait plus de code pour le codeur qui regarde plusieurs années mais bon il est plus armé que l'utilisateur/contributeur novice. Et en fait pour l'écrire c'est aussi plus facile de faire la formule d'une année à la fois parce que cela évite d'avoir toutes les feuilles d'impôt de toutes les années pour vérifier que l'on ne fait pas foirer la formule. Pour la complétude historique, nous avions souvent envisagé de mettre des helper qui calculent à part certaines expressions (ici ce serait `somme_reparations`) mais on essayé de les limiter au maximum parce que cela compliquait la lecture de l'utilisateur/contributeur novice

 - Ton 3e point est un mois  (-) pour moi. Il viole le principe _explicit is better than implicit_. Je désire que cela pète ou au moins produise un warning si la case n'existe pas à la période considérée. Et du coup je règle ton unique point (-)  
  
- Enfin, je ne pense pas que cette PR introduit le problème. Essayons de rattraper notre retard sans détériorer le code mais n'essayons pas de tout régler en un seul coup. Mais bien sûr il est très utile de relever les problèmes potentiels. Je propose que l'on répertorie ce genre de problème pour établir de guidelines consensuelles et après on fera un passage de nettoyage. Ce sera un bon sujet à mettre à l'ordre du jour de la réunion du bureau de l'association ;-)

Reply:
OK pour garder les multiples formules. Je pense qu'il reste intéressant de mettre un commentaire tel que `# report année 2009` pour facilement identifier ce qui a changé entre les formules.

Reply:
Ok pour les commentaires.

Reply:
Est-ce que ça vous va comme ça ? (J'ai pensé que ça faisait la même chose que de rajouter des commentaires mais en moins lourd, mais je peux rechanger) : 
```
def formula_2016(foyer_fiscal, period, parameters):
        '''
        Dépenses de grosses réparations des nus-propriétaires
        '''
        f6cb = foyer_fiscal('f6cb', period)
        report_2009 = foyer_fiscal('f6hj', period)
        report_2010 = foyer_fiscal('f6hk', period)
        report_2011 = foyer_fiscal('f6hl', period)
        report_2012 = foyer_fiscal('f6hm', period)
        report_2013 = foyer_fiscal('f6hn', period)
        report_2014 = foyer_fiscal('f6ho', period)
        report_2015 = foyer_fiscal('f6hp', period)
        plafond_grosses_reparations = parameters(period).impot_revenu.charges_deductibles.grosses_reparations.plafond

        return min_(f6cb + report_2009 + report_2010 + report_2011 + report_2012 + report_2013 + report_2014 + report_2015, plafond_grosses_reparations)
```